### PR TITLE
Fix off-by-one error in HexagonalGridPositions.C

### LINKED ISSUE #27051: Potential off-by-one error in HexagonalGridPositions.C
### Bug Description

Recently, #26935 added the capability to generate various positions within a Hexagonal Lattice (which I am very grateful for). I went to try it out today and encountered
```
*** ERROR ***
/home/lgross/gcmr/mwes/small_inf_assembly/coolant_channel_generator.i:21: (Positions/hex_grid/lattice_flat_to_flat):
    Bundle pitch is too small to fit this many rings at that pin pitch
```
but I think either I am misunderstanding `nr` or there is an off-by-one error. 

When counting rings, does the central ring count or not? For example, I would say this hexagaonal lattice has 6 rings
6_rings_hexlatticie

I would say that the flat-to-flat here is $5pitch\sqrt{3}$. Currently, https://github.com/idaholab/moose/blob/659e12791ceaa627c75956d97d009b3f9f8963c0/modules/reactor/src/positions/HexagonalGridPositions.C#L61 would throw an error because $6pitch\sqrt{3} > 5pitch\sqrt{3}$. I believe the correct condition to throw an error should be $(nr-1)p\sqrt{3} > lattice$ $flat$  $to$  $flat$

I've attached some diagrams for a hexagon of arbitrary number of rings `nr`. Here a hexagon is shown where the edge length is $(nr-1)p$
hex_w_tris copy
Here is the 30-60-90 triangle created that relates the side length to the flat to flat distance. Note the base of this triangle is half the flat to flat distance. 
30-60-90-tri
Here the math tells us
$$\big( \frac{1}{2} dftf \big)^2 +  \big( \frac{1}{2} (nr-1)p \big)^2 = \big((nr-1)p \big)^2$$
$$\big(\frac{1}{2} dftf \big)^2 =  \big((nr-1)p \big)^2 -  \big( \frac{1}{2} (nr-1)p \big)^2$$
$$\frac{1}{2} dftf  =  (nr-1)p \frac{\sqrt{3}}{2}$$
Thus, the minimum flat to flat distance is given by $$dftf  =  (nr-1)p \sqrt{3}$$

### Steps to Reproduce

Run this file `coolant_channel_generator.i`
```
pin_lattice_pitch = 0.02
[GlobalParams]
  quad_center_elements = true
[]

[Mesh]
  [cmg]
    type = CartesianMeshGenerator
    dx = 1
    dim = 1
  []
[]

[Positions]
  active = 'hex_grid'
  [hex_grid]
    type = HexagonalGridPositions
    center = '0.0 0.0 0.0'
    nr = 3
    pin_pitch = ${pin_lattice_pitch}
    lattice_flat_to_flat = ${fparse pin_lattice_pitch*sqrt(3)*2}
    pattern =  '0 1 0;
               1 0 0 1;
              0 0 1 0 2;
               1 0 0 1;
                0 1 0'
    include_in_pattern = '1'
    outputs = 'out'
  []
[]

# The following content is adding postprocessor(s) to check sideset areas.
# The reactor module is unfortunately quite brittle in its assignment of sideset
# IDs, so we want to be extra sure that any changes to sideset numbering are detected
# in our test suite.
[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  [out]
    type = JSON
    execute_on = 'FINAL'
    file_base = 'coolant_channel_positions'
  []
[]
```

### Impact

This would be a very small PR and I could likely get around it by using a different `lattice_flat_to_flat`, but I think it should be resolved.

### [Optional] Diagnostics

_No response_

### PR DESCRIPTION
**Pull request by lewisgross1296**
Closes #27051

## Reason
Corrects the formula for the minimum lattice flat-to-flat distance

## Design
Updated the `paramError` condition and error message.

## Impact
Fix bug

### CODE REVIEW

**Comment by lewisgross1296** [comment]
We should probably add a test that triggers this error, though I'm not the most familiar with creating MOOSE tests. If someone has a test for a similar functionality, I can make an attempt based from it and push

File: modules/reactor/src/positions/HexagonalGridPositions.C [hunk 1]
```diff
@@ -58,9 +58,10 @@ HexagonalGridPositions::HexagonalGridPositions(const InputParameters & parameter
         std::set<unsigned int>(getParam<std::vector<unsigned int>>("include_in_pattern").begin(),
                                getParam<std::vector<unsigned int>>("include_in_pattern").end()))
 {
-  if (MooseUtils::absoluteFuzzyGreaterThan(_nr * _pin_pitch * sqrt(3), _lattice_flat_to_flat))
+  if (MooseUtils::absoluteFuzzyGreaterThan((_nr - 1) * _pin_pitch * sqrt(3), _lattice_flat_to_flat))
```
Comment (GiudGiud) [inline modules/reactor/src/positions/HexagonalGridPositions.C:61]: so if you have 1 ring, then lattice_flat_to_flat should be > 0 ?
it seems 1 * sqrt(3) * pin_pitch worked better

Comment (lewisgross1296) [reply modules/reactor/src/positions/HexagonalGridPositions.C:61]: That is a good point. Though now thinking about it, maybe 1 ring is more like a special case? Maybe we check for `_nr==1` (where the flat to flat has to be at least the pin pitch). This doesn't seem to follow the formula for when `nr>1`. (i.e. when nr=0, there isn't a sqrt(3) factor needed).

Comment (GiudGiud) [reply modules/reactor/src/positions/HexagonalGridPositions.C:61]: I dont think so?
let s just rederive the formula

Comment (lewisgross1296) [reply modules/reactor/src/positions/HexagonalGridPositions.C:61]: I'm not quite sure I know what you mean. I guess it also depends on if you want it to work for `nr` as integers only. E.g. the example in the issue counting as 6 or 5.5.

Let's say `nr=1`, how would you define the minimum flat to flat?

Comment (lewisgross1296) [reply modules/reactor/src/positions/HexagonalGridPositions.C:61]: I tried with the logic we discussed (separate `nr=1`), but ran into this error
```
*** ERROR ***
Specified bundle pitch 0.069282 is too small to fit the given pins and wire wrap!
```
from `HexagonalLatticieUtils.C`.

What exactly are `wire_diameter` and `wire_pitch`? They show up here
```
HexagonalLatticeUtils::HexagonalLatticeUtils(const Real bundle_inner_flat_to_flat,
                                             const Real pin_pitch,
                                             const Real pin_diameter,
                                             const Real wire_diameter,
                                             const Real wire_pitch,
                                             const unsigned int n_rings,
                                             const unsigned int axis)
```
They have fixed values of `0.` and `1.` inside `HexagonalGridPositions::initialize()`

**Comment by lewisgross1296** [comment]
Here's the script I mentioned for local testing. The most recent commit compiles, but throws an error when running this example
```
!include common_input.i
[GlobalParams]
  quad_center_elements = true
[]

[Mesh]
  [cmg]
    type = CartesianMeshGenerator
    dx = 1
    dim = 1
  []
[]

[Positions]
  active = 'hex_grid'
  [hex_grid]
    type = HexagonalGridPositions
    center = '0.0 0.0 0.0'
    nr = 3
    pin_pitch = ${pin_lattice_pitch}
    lattice_flat_to_flat = ${fparse pin_lattice_pitch*sqrt(3)*2}
    pattern =  '0 1 0;
               1 0 0 1;
              0 0 1 0 2;
               1 0 0 1;
                0 1 0'
    include_in_pattern = '1'
    outputs = 'out'
  []
[]

# The following content is adding postprocessor(s) to check sideset areas.
# The reactor module is unfortunately quite brittle in its assignment of sideset
# IDs, so we want to be extra sure that any changes to sideset numbering are detected
# in our test suite.
[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  [out]
    type = JSON
    execute_on = 'FINAL'
    file_base = 'coolant_channel_positions'
  []
[]
```
See this [comment](https://github.com/idaholab/moose/pull/27053#discussion_r1520755821)

**Comment by GiudGiud** [comment]
without having checked the code, what I expect is that when you specify nr  of 6 for example with a flat to flat for 5.5, the lattice utils is creating a lattice of 6 rings with the smaller flat to flat.

this does not work because there are some internal checks on dimensions that are not codes for half numbers of rings.
you ll need to rework deeper into the util if you want to have a half ring

**Comment by GiudGiud** [comment]
I suspect you could just work with 6 rings instead of 5.5 and use the flat to flat for 6 rings instead of the 5.5 flat to flat (so a little bigger than what you have in your mesh).
it's ok if the mesh division extends beyond your mesh

**Comment by lewisgross1296** [comment]
Yeah since I only care about the center positions, the larger flat to flat worked great. I  ran locally with `lattice_flat_to_flat = 10`, and it output the positions for the compacts I care about.

I would love to have support for half rings, but I don't expect to have much time (for at least the next few months) to make the utility more robust. How would you like to proceed with this PR?

I do think the case checking for `nr==1` is worth merging.  This minimum flat to flat (`(_nr-q)*_pin_pitch*sqrt(3)`) is correct for half rings but is too small for full rings. I'm not entirely sure I understand the wire wrap component, but I think we could merge this if the formula was updated for integer rings only.

**Comment by GiudGiud** [comment]
Feel free to reduce the PR to just that change and add a test and some comments as to why it s the way it is.

File: modules/reactor/src/positions/HexagonalGridPositions.C [hunk 2]
```diff
@@ -58,9 +58,20 @@ HexagonalGridPositions::HexagonalGridPositions(const InputParameters & parameter
...
+  if (_nr == 1)
+  {
+    if (MooseUtils::absoluteFuzzyGreaterThan(_pin_pitch, _lattice_flat_to_flat))
+      paramError("lattice_flat_to_flat",
+                 "For one ring, the lattice flat to flat must be at least the pin pitch");
+  }
+  else
+  {
+    if (MooseUtils::absoluteFuzzyGreaterThan((_nr - 1) * _pin_pitch * sqrt(3),
+                                             _lattice_flat_to_flat))
```
Comment (GiudGiud) [inline modules/reactor/src/positions/HexagonalGridPositions.C:70]: so from your testing, specifying a flat to flat that is smaller than `nr * pitch * sqrt(3)`  breaks the utility right?
so until the utility is reworked to allow half pins, it should error when the flat to flat is below that number

I know the utility still works for cases with 5.5 pins but only with specifying:
nr = 6
flat to flat. = flat to flat from 6 rings (not the current flat to flat of 5.5 pins)

**Review by GiudGiud — DISMISSED** [review]
thanks for the fix!

**Comment by lewisgross1296** [comment]
I'm actually not convinced the code is right as is. It's been a sec, but IIRC the decision was to not support half rings yet. So then we should require that minimum flat-to-flat is `_nr*_pin_pitch*sqrt(3)`?

**Comment by GiudGiud** [comment]
you're right.
I drew it and convinced myself nr-1 was enough

you have a fix in there for the 1 pin case. the >1 case should be nr * sqrt(3)

**Comment by lewisgross1296** [comment]
> you're right. I drew it and convinced myself nr-1 was enough but it s not
> 
> you have a fix in there for the 1 pin case. the >1 case should be nr * sqrt(3)

Yeah, I'm working on a diagram to convince myself as well, but I can double check this and push the correction soon. (tomorrow likely)

**Comment by lewisgross1296** [comment]
After drawing some more pictures, I think the pattern is like this
| nr | dftf minimum|
| ------------- | ------------- |
| $1$ | $pitch$ |
| $2$ | $5r$ |
| $3$ | $8r$ | 
| $4$ | $11r$ | 
| ... | ... |
|$nr$ | $2*(1 + (nr-1)*1.5)*r$ |


where $r$ is the radius of a hexagon (i.e. center to vertex distance) and we known that $pitch = \sqrt{3} r$.

I can upload some pictures if desired (they're a bit crude though)

**Comment by GiudGiud** [comment]
This is right
so instead of nr * pitch * sqrt(3) = nr * 3 * r
we should have
(nr * 3 - 1) * r

**Comment by lewisgross1296** [comment]
I should add some tests too for this, but I prob won't get a chance until tonight or later this week.

**Review by GiudGiud — APPROVED** [review]
Test failure is unrelated cancelation
my only contribution is adding two exception tests so it counts as "minor"

Thanks for the contribution @lewisgross1296

**Comment by lewisgross1296** [comment]
Thanks for finishing this up!

**Pull request merged**